### ISSUE TITLE
3d spatialization

### ISSUE THREAD

**Issue by Tetsujinfr** [SoundScapeRenderer/ssr#201]
hi, do you have any plan to support 3d spatialization at some point?

**Comment by mgeier** [comment]
The latest development version (in the `master` branch) already has some 3D features, but not all parts of the SSR support 3D yet.

For now, `ssr-binaural` is the only renderer that supports 3D, and only when using SOFA files (see #165).

The Qt-based GUI, however, does only show a 2D top view (and there are currently no plans to extend that).

There is a new (and  very experimental) 3D browser-based GUI, see https://ssr.readthedocs.io/en/latest/network-websocket.html.

The new network interfaces (WebSocket and FUDI) fully support 3D.

There are concrete plans to extend the scene description format to 3D, see #155.

Other than that, I don't know of any concrete plans, but it would of course be great to have full 3D support!

Here are some slides I made recently about the latest SSR developments: https://github.com/mgeier/fk2019

**Comment by Tetsujinfr** [comment]
Thanks for your great reply and all the information. I will look at the code and see if I can hook it up with some 3D renderer to visualize the 3D spatialized sound source.
Do you have some benchmarks regarding performance of your binaural model w.r.t compute load? e.g. can it run on a small device lie a smartphone or does it require some powerful desktop cpu?

**Comment by Tetsujinfr** [comment]
Ok , so I have done many tests, with and without the jetson-clocks scripts: the camera works in both cases, so the USB 3.1 works in normal power mgmt mode or in full speed mode either. Great. Now I understand why you trigger the full speed mode, in mormal speed mgmt mode there are some significant slowdowns just when capturing the video feed via opencv, even on the NX, which is a bit surprising but I guess it is what it is.

However, I find out that after reboot, my zed mini is not detected properly and I need to re-plug it so that it gets recognized by the system properly.
That looks similar to [this issue thread](https://github.com/stereolabs/zed-ros-wrapper/issues/249).
When reading the thread it is not clear to me what the solution to this problem is. You said to use the latest jetpack 4.2 to fix the issue but 1) users have said that did not fix the problem and 2) I am using jetpack 4.4 so I would have thought it would fix the issue if jetpack 4.2 was supposed to fix it on the first place.

Any advice on fixing the "at boot ZedM camera connectivity issue"? 
thanks

**Comment by Tetsujinfr** [comment]
Please not that in normal speed mgmt mode on the NX devkit, ZedFu and ZED_Depth_Viewer have good performance (no slowdown). With Zed_Explorer, very few frames are dropped: 0.07% for WVGA@100fps, 0.2% for 720p@60fps and , 0.1% for 2K@15fps and HD@30fps (I have removed the first cple dropped frames at mode initialization which looks like outliers at startup time).
So I suspect that the perf issues I observed in my previous post were due to poorly optimized python code using opencv.

Still have the USB connection issue at boot, so I need to unplug/plug the camera after each reboot, painful.

**Comment by mgeier** [comment]
> I will look at the code and see if I can hook it up with some 3D renderer to visualize the 3D spatialized sound source.

You should check out the above-mentioned browser GUI. Even though it's very bare-bones, it can already visualize 3D sources.

Feel free to make improvements to the browser GUI, I'm looking forward to your contributions!

> Do you have some benchmarks regarding performance of your binaural model w.r.t compute load? e.g. can it run on a small device lie a smartphone or does it require some powerful desktop cpu?

There are no "official" benchmarks, but I'd say it's quite fast.
There isn't really a "binaural model", there are only impulse responses which are convolved with the input signal(s).
Therefore, the performance mostly depends on the length of the impulse responses and the number of sources.

IOW, if the impulse responses are not exceedingly long and you only have a few sources, it definitely would run on a smartphone (if you could somehow manage to compile the SSR on it, which I guess won't work).

On a desktop CPU you can typically render hundreds of sources at the same time.

Anyway, you should just try it with your impulse responses on your target CPU, that's the only thing that counts.
And please let us know about the results!

**Comment by janhenhan** [comment]
Hey Matthias, just double-checking my understanding: SSR's binaural renderer is now supporting 3D positioned sources that can be visualized/modified with the 3D GUI? You're not supporting Yaw/Pitch/Roll 3D-Headtracking yet, right? I got Ben's headtracker ("Supperware") yesterday and for the first time feel confident enough in the tracking capabilities of a low-cost tracker, such that this feature would actually become interesting. Let me know in case I'm wrong and I will update my SSR right away ;)

**Comment by mgeier** [comment]
> SSR's binaural renderer is now supporting 3D positioned sources that can be visualized/modified with the 3D GUI?

Correct. The `master` version does. And there will be a new release quite soon, I hope.

The 3D visualization and interaction is quite bare-bones, but it works.

If you want to improve it, that would be great!

> You're not supporting Yaw/Pitch/Roll 3D-Headtracking yet, right?

Correct.

The general infrastructure is in place, but the individual tracker classes haven't yet been updated to 3D.

Would you like to fix this?

Related: #140, #146.

**Comment by mgeier** [comment]
PS: if you manage to create network messages from your tracker that are compatible with one of the network interfaces (https://ssr.readthedocs.io/en/latest/network.html), you could use that for 3D tracking, because the network interfaces already support 3D orientations.

**Comment by janhenhan** [comment]
Awesome - your P.S. would have been my approach anyway! That's great news then, I'll look into it.
